frontend/packages/pipelines-plugin/src/components/pipelines/pipeline-topology/utils.ts

```diff
@@ -223,14 +229,87 @@ export const getEdgesFromNodes = (nodes: PipelineMixedNodeModel[]): PipelineEdge
     }),
   ).filter((edgeList) => !!edgeList);
 
+export const getFinallyTaskHeight = (allTasksLength: number, disableBuilder: boolean): number => {
+  const FINALLY_TASK_NODE_HEIGHT = NODE_HEIGHT + FINALLY_NODE_VERTICAL_SPACING;
+  return (
+    NODE_HEIGHT +
+    (allTasksLength - 1) * FINALLY_TASK_NODE_HEIGHT +
+    (!disableBuilder ? NODE_HEIGHT : 0) +
+    FINALLY_NODE_PADDING * 2
+  );
+};
+
+export const getLastRegularTasks = (nodes: PipelineMixedNodeModel[]): string[] => {
+  const regularTasks = nodes;
+  const runAfters: string[] = regularTasks.reduce((acc, regularTask) => {
+    const {
+      data: {
+        task: { runAfter },
+      },
+    } = regularTask;
+    if (runAfter && runAfter.length) {
+      runAfter.forEach((ra) => {
+        if (!acc.includes(ra)) {
+          acc.push(ra);
+        }
+      });
+    }
+    return acc;
+  }, []);
+
+  return _.difference(
+    regularTasks.map((n) => n.id),
+    runAfters,
+  );
+};
+
+export const finallyTasksToNodes = (
+  nodes: PipelineMixedNodeModel[],
+  pipeline?: PipelineKind,
+  pipelineRun?: PipelineRunKind,
+): PipelineMixedNodeModel[] => {
+  if (!pipeline.spec?.finally) {
+    return nodes;
+  }
+  const finallyTasks = pipeline.spec.finally;
+  const regularRunAfters = getLastRegularTasks(nodes);
+  const name = 'finally-node';
+  const finallyGroupNode = createFinallyNode(getFinallyTaskHeight(finallyTasks.length, true))(
+    name,
+    {
+      disableBuilder: true,
+      isFinallyTask: true,
+      pipeline,
+      pipelineRun,
+      task: {
+        name,
+        runAfter: regularRunAfters,
+        finallyTasks: finallyTasks.map((ft) => ({
+          ...ft,
+          disableTooltip: false,
+        })),
+      },
+    },
+  );
+  return [finallyGroupNode];
+};
+
 export const getTopologyNodesEdges = (
   pipeline: PipelineKind,
   pipelineRun?: PipelineRunKind,
 ): { nodes: PipelineMixedNodeModel[]; edges: PipelineEdgeModel[] } => {
   const taskList: PipelineVisualizationTaskItem[] = _.flatten(
     getPipelineTasks(pipeline, pipelineRun),
   );
-  const nodes: PipelineMixedNodeModel[] = tasksToNodes(taskList, pipeline, pipelineRun);
+  const taskNodes: PipelineMixedNodeModel[] = tasksToNodes(taskList, pipeline, pipelineRun);
+
+  const finallyTaskNodes: PipelineMixedNodeModel[] = finallyTasksToNodes(
+    taskNodes,
+    pipeline,
+    pipelineRun,
+  );
```

Comment:
Curious, why did you over-reach and do [ODC-4312](https://issues.redhat.com/browse/ODC-4312)? @nemesis09 FYI.

Reply:
Unlike pipeline builder, visualization in details page will not always have `finally section` (depending on pipeline definition),  so i had to conditionally handle the graph width and height based on the finally task availability, during the process I had to add this. But there are some more things that needs to handled in ODC-4312 like `Finally links` , `Finally task status`, `Tooltips` apart from this shared visualization component.

Reply:
Not sure you actually had to do this at all... since the graph would not need resizing if it did not render the finally outside of the builder. But it is what it is. Thanks for the clarification.